halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/services/v1/GenerateService.java

```diff
@@ -123,45 +124,57 @@ public ResolvedConfiguration generateConfig(String deploymentName, List<String>
-        profile.writeStagedFile(spinnakerStagingPath);
-      }
-
-      Map<String, Profile> profileMap = new HashMap<>();
-      for (Profile profile : profiles) {
-        profileMap.put(profile.getName(), profile);
-      }
+      Map<String, Profile> outputProfiles = processProfiles(profiles);
 
-      serviceProfiles.put(service.getType(), profileMap);
-    }
+      List<Profile> customProfiles = userProfileNames.stream()
+          .map(s -> (Optional<Profile>) service.customProfile(deploymentConfiguration, runtimeSettings, Paths.get(userProfilePath.toString(), s), s))
+          .filter(Optional::isPresent)
+          .map(Optional::get)
+          .collect(Collectors.toList());
 
-    // Step 3.
-    Path userProfilePath = halconfigDirectoryStructure.getUserProfilePath(deploymentName);
+      pluralModifier = customProfiles.size() == 1 ? "" : "s";
+      DaemonTaskHandler.message("Discovered " + customProfiles.size() + " custom profile" + pluralModifier + " for " + service.getCanonicalName());
+      outputProfiles.putAll(processProfiles(customProfiles));
 
-    if (Files.isDirectory(userProfilePath)) {
-      DaemonTaskHandler.newStage("Copying user-provided profiles");
-      File[] files = userProfilePath.toFile().listFiles();
-      if (files == null) {
-        files = new File[0];
-      }
-
-      Arrays.stream(files).forEach(f -> {
-        try {
-          DaemonTaskHandler.message("Copying existing profile " + f.getName());
-          Files.copy(f.toPath(), Paths.get(spinnakerStaging.toString(), f.getName()), StandardCopyOption.REPLACE_EXISTING);
-        } catch (IOException e) {
-          throw new HalException(
-              new ConfigProblemBuilder(Severity.FATAL, "Unable to copy profile \"" + f.getName() + "\": " + e.getMessage() + ".").build()
-          );
-        }
-      });
+      serviceProfiles.put(service.getType(), outputProfiles);
     }
 
     return new ResolvedConfiguration()
         .setServiceProfiles(serviceProfiles)
         .setRuntimeSettings(runtimeSettings);
   }
 
+  private Map<String, Profile> processProfiles(List<Profile> profiles) {
+    for (Profile profile : profiles) {
+      profile.writeStagedFile(spinnakerStagingPath);
+    }
+
+    Map<String, Profile> profileMap = new HashMap<>();
+    for (Profile profile : profiles) {
+      profileMap.put(profile.getName(), profile);
+    }
+
+    return profileMap;
+  }
+
+
+  private static List<String> aggregateProfilesInPath(String basePath, String relativePath) {
+    String filePrefix;
+    if (!relativePath.isEmpty()) {
+      filePrefix = relativePath + File.separator;
+    } else {
+      filePrefix = relativePath;
+    }
+
+    File currentPath = new File(basePath, relativePath);
+    return Arrays.stream(currentPath.listFiles())
+        .map(f -> f.isFile() ? Collections.singletonList(filePrefix + f.getName()) : aggregateProfilesInPath(basePath, filePrefix + f.getName()))
+        .reduce(new ArrayList<>(), (a, b) -> {
```

Comment:
`a, b`? this is just a standard reducer, I don't know what else to name them... `left, right`? `one, two`?